### PR TITLE
APEXMALHAR-2514 JDBC poll offset rebase support.

### PR DESCRIPTION
**Pull request by tweise**
based on #649

### CODE REVIEW

**Comment by tweise** [comment]
@sandeshh would you like to review this PR?

**Comment by tweise** [comment]
@bhupeshchawda can you review?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 1]
```diff
@@ -136,13 +139,24 @@
...
   protected boolean isPollerPartition;
 
-  protected transient MutablePair<Integer, Integer> currentWindowRecoveryState;
-
   private transient int lastOffset;
+  private transient Object prevKey;
+  private transient Object lastKey;
+
+  /**
+   * The candidate key/offset pair identified by the fetch thread
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Suggest using "poller thread"  instead of "fetch thread" since the thread class is called "DBPoller", reader will be able correlate the two more easily.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 2]
```diff
@@ -750,6 +813,24 @@ public void setResultLimit(int resultLimit)
...
   }
 
-  private static final Logger LOG = LoggerFactory.getLogger(AbstractJdbcPollInputOperator.class);
+  public boolean isRebaseOffset()
+  {
+    return rebaseOffset;
+  }
+
+  /**
+   * Whether the query should automatically be be augmented with a WHERE
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: typo: be be => be

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 3]
```diff
@@ -246,6 +275,25 @@ public void emitTuples()
...
   protected abstract void emitTuple(T tuple);
 
+  /**
+   * Visible to subclasses to allow for custom offset saving and initialization.
+   */
+  protected static class WindowData
+  {
+    public Object key;
+    public int lowerBound;
+    public int upperBound;
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Can these fields be made protected ?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 4]
```diff
@@ -423,11 +466,15 @@ public void emitReplayedTuples(PreparedStatement ps)
...
+    final List<KeyValPair<Integer, Integer>> partitionRanges;
     try {
       store.connect();
       dslContext = createDSLContext();
-      partitionToRangeMap = getPartitionedQueryRangeMap(getPartitionCount());
+      if (getPartitionCount() > 0) {
+        partitionRanges = getPartitionedQueryRanges(getPartitionCount());
+      } else {
+        partitionRanges = new ArrayList<>(0);
+      }
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Looks better if this check is moved into getPartitionedQueryRanges; also makes that function more robust by avoiding divide-by-zero errors in future maintenance changes.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 5]
```diff
@@ -464,52 +512,66 @@ public void partitioned(
...
     int rowCount = 0;
     try {
-      rowCount = getRecordsCount();
+      rowCount = getRecordsCount(null);
     } catch (SQLException e) {
       LOG.error("Exception in getting the record range", e);
     }
 
-    HashMap<Integer, KeyValPair<Integer, Integer>> partitionToQueryMap = new HashMap<>();
+    List<KeyValPair<Integer, Integer>> partitionToQueryMap = new ArrayList<>();
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: partitionToQueryMap => partitionToQueryList

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 6]
```diff
@@ -282,23 +330,44 @@ public void deactivate()
...
+  protected int insertDbDataInQueue(PreparedStatement preparedStatement) throws SQLException, InterruptedException
   {
+    int resultCount = 0;
     preparedStatement.setFetchSize(getFetchSize());
     ResultSet result = preparedStatement.executeQuery();
     if (result.next()) {
       do {
-        while (execute && !emitQueue.offer(getTuple(result))) {
+        T obj = getTuple(result);
+        while (execute && !emitQueue.offer(obj)) {
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Not part of this PR but seems like emitQueue does not need to be a LinkedBlockingDeque since we are not using the deque feature of adding at both ends or removing from both.

**Comment by amberarrow** [comment]
Made a few comments; once addressed, merge can proceed

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 7]
```diff
@@ -282,23 +330,44 @@ public void deactivate()
...
-  protected void insertDbDataInQueue(PreparedStatement preparedStatement) throws SQLException, InterruptedException
+  protected int insertDbDataInQueue(PreparedStatement preparedStatement) throws SQLException, InterruptedException
   {
+    int resultCount = 0;
     preparedStatement.setFetchSize(getFetchSize());
     ResultSet result = preparedStatement.executeQuery();
     if (result.next()) {
       do {
-        while (execute && !emitQueue.offer(getTuple(result))) {
+        T obj = getTuple(result);
```
Comment (amberarrow) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Can we check the result for null and skip it ? Then we won't need to null check inside the loop in emitTuples().

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 6 above)
Comment (tweise) [reply library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: I modified the loop as there is no reason to check for result.next() first and then have a do .. while.

**Comment by tweise** [comment]
@amberarrow comments are addressed. Please mark the PR approved or merge it.

**Comment by amberarrow** [comment]
Approved.
